File: leetcode/206_reverse_linked_list.cpp
using namespace std; 

#include <vector>

class Solution {

struct ListNode {
    int val;
    ListNode *next;
    ListNode(int x) : val(x), next(NULL) {}
};

public:
    ListNode* reverseList(ListNode* head) {
        ListNode* prev = NULL; 
        ListNode* curr = head; 
        ListNode* next = NULL; 
        
        while(curr != NULL) {
            next = curr->next; 
            curr->next = prev; 
            prev = curr; 
            curr = next; 
        }
        
        return prev; 
    }
};

File: leetcode/topics/listnode/206_reverse_linked_list.go
package listnode

func reverseList(head *ListNode) *ListNode {
	if head == nil || head.Next == nil {
		return head
	}

	//     var prev, next *ListNode
	//     curr := head
	//     for curr != nil {
	//         next = curr.Next
	//         curr.Next = prev
	//         prev = curr
	//         curr = next
	//     }
	//     return prev

	newHead := reverseList(head.Next)
	head.Next.Next = head
	head.Next = nil

	return newHead
}
